rfcs/20220201-quantization-scaling/README.md

```diff
@@ -0,0 +1,264 @@
+| LRN                       |                             |
+| (Log)Softmax              | attr::output_scale[[#2]][2] |
+| Pooling                   | attr::p_op::binary_mul      |
+| PRelu                     |                             |
+| Resampling                | attr::p_op::binary_mul      |
+| Shuffle                   |                             |
+| Sum                       |                             |
+| attr::p_op::sum           | built-in parameter          |
+| Reorder                   | attr::output_scale          |
+| Reduction                 | attr::p_op::binary_mul      |
+
+
+## 2. Addressing short-term requests
+
+### 2.1. Run-time scaling in Eltwise
+
+This feature doesn't require API extensions, because scaling can be implemented
+as a separate Binary post-op.
+
+
+### 2.2. Scaling for src1 in Binary post-op
+
+#### 2.2.1 Option 1
+
+New API is required to support the requested feature. New API extends binary
+post-op with a `scale` parameter (scaling mechanism `built-in parameter`). This
+mechanism is aligned with scaling for other post-operations and shares the same
+limitations:
+- only single-value scaling;
+- no run-time support.
+
+API:
+
+```c
+dnnl_status_t DNNL_API dnnl_post_ops_append_binary_v2(dnnl_post_ops_t post_ops,
+        float scale, dnnl_alg_kind_t alg_kind,
+        const dnnl_memory_desc_t *src1_desc);
+
+dnnl_status_t DNNL_API dnnl_post_ops_get_params_binary_v2(
+        const_dnnl_post_ops_t post_ops, int index, float *scale,
+        dnnl_alg_kind_t *alg_kind, const dnnl_memory_desc_t **src1_desc);
+```
+
+An example:
+
+```cpp
+    post_ops p_ops;
+    p_ops.append_binary(scale, algorithm::binary_add, summand_md);
+
+    primitive_attr attr;
+    attr.set_post_ops(p_ops);
+```
+
+#### 2.2.2 Option 2
+
+An alternative solution would be to extend mechanism `attr::scales`. This
+solution doesn't require new API, but extends the existing one.
+
+An example:
+
+```cpp
+    post_ops p_ops;
+    p_ops.append_binary(algorithm::binary_add, summand_md);
+
+    primitive_attr attr;
+    attr.set_post_ops(p_ops);
+    attr.set_scales(DNNL_ARG_ATTR_MULTIPLE_POST_OP(0) | DNNL_ARG_SRC_1,
+            /* mask */ 0, {DNNL_RUNTIME_F32_VAL});
+```
+
+While the alternative solution allows to avoid new API, it introduces one more
+way to set scale which complicates both implementation and integration.
+Please review section
+[Unifying scaling in long-term](#3-unifying-scaling-in-long-term) for more
+details.
+
+
+### 2.3. Scales for inputs in Concat
+
+The mechanism `attr:set_scales` will be introduced to Concat primitive. This
+mechanism is used in Binary primitive, so API is already there.
+
+The following limitations are in place for an initial implementation:
+ - no-runtime support;
+ - mask is 0.
+
+An example:
+
+```cpp
+    primitive_attr attr;
+    for (int n = 0; n < num_src; ++n)
+        attr.set_scales(DNNL_ARG_MULTIPLE_SRC + n, /* mask */ 0, src_scales[n]);
```

Comment:
if we already have `mask` in the function parameter list, why is it 0 only (mentioned at L177)? An implementation limitation?

Reply:
It is an implementation limitation. Also it is an example, so I tried to simplify it. Since API supports mask we will be able to extend support of mask for concat if it is requested by users.